## Limited SARS-CoV-2 diversity within hosts and following passage in cell culture

Authors: Gage K. Moreno1*, Katarina M. Braun2*, Peter J. Halfmann2, Trent M. Prall1, Yoshihiro Kawaoka2, Thomas C. Friedrich2,3, David H. O’Connor1,3 

\* **These authors contributed equally to this work**  

1. Department of Pathology and Laboratory Medicine, University of Wisconsin-Madison, Madison, WI, United States of America
2. Department of Pathobiological Sciences, University of Wisconsin-Madison, Madison, WI, United States of America
3. Wisconsin National Primate Research Center, University of Wisconsin-Madison, Madison, WI, United States of America

**Abstract**:

Since the first reports of a novel pneumonia-causing respiratory virus (COVID-19) emerged in Wuhan, Hubei province, China, there have been considerable efforts to sequence the causative virus, SARS-CoV-2 (also referred to as hCoV-19) and to make viral genomic information available quickly on shared repositories. As of 30 March 2020, 2,765 consensus sequences have been shared on GISAID, the principal repository for SARS-CoV-2 genetic information. These sequences are primarily consensus sequences from clinical and passaged samples, but few reports have looked at diversity within virus populations. Understanding such diversity is essential to understand viral evolutionary dynamics. Here, we characterize within-host viral diversity from primary isolates and passaged samples, all originally deriving from an individual returning from Wuhan, China, who was diagnosed with COVID-19 and subsequently sampled in Wisconsin, United States. We use a metagenomic approach with Oxford Nanopore Technology (ONT) in combination with Illumina Miseq to capture minor frequency variants ≥1%. In a clinical swab obtained from the day of hospital admission, we identify 15 single nucleotide variants (SNVs) ≥1% frequency, primarily located in ORF1a. While diversity is low overall, the dominant viral genetic signatures are likely secondary to population size changes, with some evidence for diversifying selection in ORF1a. We see little to no evidence of positive selection or ongoing adaptation of SARS-CoV-2 to human hosts in the limited samples available to us. 

**What is contained in this repository**: 

[to-do]

**How to use this repository**: 

[to-do]

**Contacts if you have questions:**

[to-do]
