# Single nucleotide variants (SNVs)

**Author**: Katarina Braun 

## Overview

This script cleans the VCF files generated by the `Sniffles` pipeline and generates SNV plots as well. A version of this SNV plot is described as figure 3 in the manuscript.

Additionally, at the end of this notebook we plot SNVs that are shared across the primary swab as well as P1 and P2 (`SNVs-over-time.pdf`). 

## Dependencies 
- pandas
- numpy
- matplotlib
- glob
- os
- functools
- pandas
- itertools
- glob
- os
- sklearn 
- pysam
- random 
- seaborn 

**Visualize**: 
One scatterplot per gene where: 
- x axis contains gene length 
- y axis has FREQ 
- datapoints are plotted as closed if nonsynonymous and open if synonymous

### a note about numbering schemes 

Everything was mapped to the Madison, WI case reference sequence: MT039887

The numbering schemes in input VCFs differs between the Illumina and ONT files. 
1. **Illumina**: starts from 1 at the beginning of each ORF. 
2. **ONT**: starts from 1 at the beginning of the reference file and increases sequentially through the entire genome. Some of the code below will adjust the numbering scheme in the ONT files to match the Illumina files -- but just make sure you are aware of this if you are looking at the upstream VCF files. 

## Input: 

1. SNV files in VCF format (output from `Sniffles`) located here: 
    - `SARSCoV2_passage_MS/figures/*` (6 files)
    
2. SNV files in VCF format (output from the ONT pipeline) located here: 
    - `SARSCoV2_passage_MS/data_derived/figure3_SNVs/VCFs_ONT/*` (6 samples) 

## Output: 

1. Cleaned generated from VCF input files will be located here:       
    - `SARSCoV2_passage_MS/data_derived/figure3_SNVs/VCFs_Illumina/cleaned/*`

2. Spliced data by sample ID, gene, and synonymous vs nonsynonymous will be located here:   
    - `SARSCoV2_passage_MS/data_derived/figure3_SNVs/VCFs_Illumina/by_gene/*`

3. One plot for index SNPs and one plot for contact SNPs will be found here:   
    - `SARSCoV2_passage_MS/data_derived/figures/figure3.pdf`
    
## Before running this notebook ...: 

It is a good idea to update matplotlib from the command line: `conda update matplotlib`
    
